S3 파일 업로드 기능 추가

c0smosaur opened the pull request:
## #️⃣연관된 이슈

#5 

## 📝작업 내용

- S3 클라이언트 설정, 파일 업로더 클래스를 추가하였습니다.
  - `S3Config`에서 AWS 액세스키, 리전, 버킷 이름의 속성값을 받아 S3 클라이언트 빈을 생성합니다. 
  - `S3Uploder`에 파일 업로드, 파일 url 조회, 파일 삭제 기능을 추가하였습니다. 
- Mockito로 유닛 테스트를 작성하였습니다. 
  - 테스트 실행에는 `logback-test.xml`(기본적인 로깅만)을 추가하여 사용하였습니다. 

### 스크린샷 (선택)

## 💬리뷰 요구사항(선택)

- final 키워드 때문에 생성자 주입 방식으로 작성하였는데 만약 통일성을 해치거나 한다면 변경하겠습니다!
- 레포지토리를 포크해서 사용하고 있는데, 필요없는 커밋 단계가 늘어난다거나 싶으시면 말씀해주세요!
- 테스트 작성이 미숙하여 오류가 많을 수 있습니다... 
  - `logback-spring.xml`을 사용하려 하니 경로나 환경변수가 입력이 되어 있는데도 오류가 나서 디폴트 이름으로 xml 파일을 하나 추가하여 사용하였습니다. 제대로 된 해결책은 아닌 것 같아서... 더 찾아보도록 하겠습니다! ㅠㅠ

c0smosaur committed:
chore: AWS 의존성 추가

c0smosaur committed:
chore: application.yml에 AWS 설정값 추가

c0smosaur committed:
config: S3 클라이언트 생성

c0smosaur committed:
feat: S3 파일 업로드, 열람, 삭제 기능 추가

c0smosaur committed:
test: 유닛 테스트용 logback 추가

- logback-spring.xml 경로를 넣어도 계속 오류가 나서 일단 디폴트 logback-test.xml 추가

c0smosaur committed:
test: S3Uploader 유닛 테스트 추가

c0smosaur committed:
Merge branch 'teamCleanie:main' into feature/241027-#5-s3-file-upload

c0smosaur committed:
chore: application.yml 속성값 주석 제거

c0smosaur committed:
refactor: 프로젝트명 변경 refab -> repatch 반영

stoneHee99 commented on src/main/java/cleanie/repatch/common/s3/config/S3Config.java:
```diff
@@ -0,0 +1,38 @@
+package cleanie.repatch.common.s3.config;
+
+import com.amazonaws.auth.AWSCredentials;
+import com.amazonaws.auth.AWSStaticCredentialsProvider;
+import com.amazonaws.auth.BasicAWSCredentials;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3ClientBuilder;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class S3Config {
+
+    private final String AWS_REGION;
+    private final String AWS_ACCESS_KEY;
+    private final String AWS_SECRET_KEY;
+
+    public S3Config(
+            @Value("${cloud.aws.region.static}") String AWS_REGION,
+            @Value("${cloud.aws.credentials.access-key}") String AWS_ACCESS_KEY,
+            @Value("${cloud.aws.credentials.secret-key}") String AWS_SECRET_KEY) {
+        this.AWS_REGION = AWS_REGION;
+        this.AWS_ACCESS_KEY = AWS_ACCESS_KEY;
+        this.AWS_SECRET_KEY = AWS_SECRET_KEY;
+    }
+
+    @Bean
+    public AmazonS3 amazonS3Client() {
+        AWSCredentials credentials = new BasicAWSCredentials(AWS_ACCESS_KEY, AWS_SECRET_KEY);
+
+        return AmazonS3ClientBuilder
+                .standard()
+                .withCredentials(new AWSStaticCredentialsProvider(credentials))
+                .withRegion(AWS_REGION)
+                .build();
+    }
+}
```
`@Value` 어노테이션을 통해서, 매번 가져오는 부분이 불편하면 `@ConfigurationProperties`를 사용하면 조금 더 간단하게 표현할 수 있는 점도 알아두시면 좋을 것 같아요! 지금 구조도 너무 좋아요 👍 

```java
@Configuration
@ConfigurationProperties(prefix = "cloud.aws")
public class S3Config {
    private final Credentials credentials;
    private final Region region;

    public S3Config(Credentials credentials, Region region) {
        this.credentials = credentials;
        this.region = region;
    }

    @Bean
    public AmazonS3 amazonS3Client() {
        AWSCredentials awsCredentials = new BasicAWSCredentials(
            credentials.getAccessKey(), 
            credentials.getSecretKey()
        );
        return AmazonS3ClientBuilder
                .standard()
                .withCredentials(new AWSStaticCredentialsProvider(awsCredentials))
                .withRegion(region.getStatic())
                .build();
    }

    @Getter
    public static class Credentials {
        private final String accessKey;
        private final String secretKey;

        public Credentials(String accessKey, String secretKey) {
            this.accessKey = accessKey;
            this.secretKey = secretKey;
        }
    }

    @Getter
    public static class Region {
        private final String static;

        public Region(String static) {
            this.static = static;
        }
    }
}
```

c0smosaur replied:
@Value 애너테이션을 대체할 방법을 알려주셔서 감사합니다!
첨부해주신 코드가 훨씬 결합도도 낮아지고, 추후 aws 설정이 늘어날 경우에 대비해서 훨씬 확장성이 좋은 것 같아요! 
참고해서 수정하도록 하겠습니다!

stoneHee99 commented on src/main/java/cleanie/repatch/common/s3/helper/S3Uploader.java:
```diff
@@ -0,0 +1,40 @@
+import java.net.URL;
+
+@Component
+public class S3Uploader {
```
현재 `S3Uploader` 클래스는 크게 AWS S3 저장소에 파일을 전송하거나 혹은 삭제하는 책임을 가지고 있어요. 그런데 Uploader 라는 이름이 개발자 입장에서 파일 업로드 만의 책임을 가지고 있다고 생각할 수도 있을 것 같아요!

`S3FileUploader` 와 `S3FileRemover` 등의 클래스로 나누어보면 어떨까요?
너무 과하게 책임을 나눈 것 같이 느껴지신다면, `S3FileManager` 등의 작명을 해보는것도 조금 더 명확하게 해당 클래스의 책임을 나타낼 수 있을 것 같아요!

c0smosaur replied:
확실히 이름이 책임에 비해 축소되어 있네요 좋은 지적 감사합니다! 
`S3FileManager`로 이름을 변경하도록 하겠습니다. 클래스의 책임이 잘 드러나도록 작명에 더 신경써야겠네요..!

stoneHee99 commented on src/main/resources/application.yml:
```diff
@@ -25,3 +25,13 @@ decorator:
+      access-key: ${AWS_ACCESS_KEY}
+      secret-key: ${AWS_SECRET_ACCESS_KEY}
+    region:
+      static: ${AWS_REGION}
```
IDE의 환경변수 기능을 사용해서 설정 값을 노출하지 않은 점 좋습니다 👍

stoneHee99 commented on src/main/java/cleanie/repatch/common/s3/helper/S3Uploader.java:
```diff
@@ -0,0 +1,40 @@
+public class S3Uploader {
+
+    private final String BUCKET_NAME;
+    private final AmazonS3 amazonS3Client;
```
적극적으로 불변(final) 키워드를 사용하는것 좋아요! 👍

stoneHee99 commented on src/test/java/cleanie/repatch/common/s3/helper/S3UploaderTest.java:
```diff
@@ -0,0 +1,70 @@
+
+@SpringBootTest
+@TestPropertySource(properties = "cloud.aws.s3.bucket=${AWS_S3_BUCKET_NAME}")
+class S3UploaderTest {
```
테스트코드 작성 너무 좋아요! 💯 💯 💯

stoneHee99 commented on src/test/java/cleanie/repatch/common/s3/helper/S3UploaderTest.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    @Test
+    void testDeleteFile() {
+        // given
+        String fileName = "uploads/test-file.jpg";
+
+        // when
+        s3Uploader.deleteFile(fileName);
+
+        // then
+        verify(amazonS3, times(1)).deleteObject(BUCKET_NAME, fileName);
+    }
```
전체적으로 모든 테스트 코드를 `given`, `when`, `then` 의 구조로 작성하신 것 같아요!
테스트 코드를 작성해보시면서 감을 잡으셨을지(?) 모르겠지만! 유지보수하기 좋은 코드를 작성하기 위해선 테스트 코드의 작성이 수반되고, 어쩔때는 실제 프로덕션 코드보다도 테스트 코드가 길어지는 경우가 태반입니다!

매번 모든 테스트 코드에 불필요한 주석이 추가되면 가독성이 심히 떨어지게 될 수 있어요! 좋은 코드를 작성하신다면 (지금처럼요!!) 저 주석이 없어도 테스트 코드 안에서 의도를 잘 나타낼 수 있을 것 같아요 👍 

given when then 으로 이어지는 구조가 절대 나쁜건 아닙니다~! 해당 구조를 나타내기위해서 지금처럼 문단을 한줄 띄어서 구분하는 방법도 있고, `BDDMockito` 등의 라이브러리를 사용하면 테스트 코드 작성에서 해당 의도를 나타낼 수도 있으니 참고 부탁드려요 :)

stoneHee99 replied:
까먹고 남기지 못한 점이 있는데 현재`Junit` 안에 있는 `Assertions`를 통해 검증을 하신것 같아요!
SpringStarter의 Test 의존성에는 `Junit` 과 `AssertJ` 총 두가지 테스트 라이브러리를 제공해요.

이 두가지 라이브러리의 차이점을 알고 검증 부분에서는 `AssertJ`를 사용하시는게 좀 더 작성하기 용이하실거에요 👍

c0smosaur replied:
테스트 코드에서 주석은 필요할 때만 작성하도록 해야겠네요!
`BDDMockito` 사용법이나, `JUnit`, `AssertJ`의 차이점에 대해서도 더 찾아보고 사용하도록 하겠습니다! 
알려주셔서 감사합니다 ☺️

stoneHee99 commented on src/main/resources/logback-test.xml:
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<configuration>
+    <appender name="console" class="ch.qos.logback.core.ConsoleAppender">
+        <encoder>
+            <pattern>%d{yyyy-MM-dd HH:mm:ss.SSS} [%thread] %-5level %logger{39} - %msg%n</pattern>
+        </encoder>
+    </appender>
+
+    <root level="INFO">
+        <appender-ref ref="console"/>
+    </root>
+</configuration>
```
테스트 코드 실행에서의 로깅을 위해 `logback-test` 를 생성하신게 맞을까요?
아니면 기존 `logback-spring` 의 사용이 잘 안되어서 임시로 만드신걸까요 🤔

만약 전자라면 테스트 코드 실행에 로깅이 왜 필요했는지도 궁금해요 !

c0smosaur replied:
테스트 코드 실행 과정에서 로그 파일 관련 오류가 발생하길래, 가장 먼저 찾는다는 이름으로 임시로 추가하였습니다. 
🥲 따로 테스트 코드 실행에 로깅 필요가 없다면, 제거하도록 하겠습니다!

stoneHee99 replied:
아니에요!! 이유가 궁금했던거지 필요한 코드라면 있어도 좋을 것 같아요 :)

stoneHee99 reviewed (approved):
전체적으로 코드를 너무 잘 짜주셨어요..!!

특히 파일 업로드 기능은 어떻게 보면 외부 의존성이 큰 기능이기 때문에 흔히 테스트하기 어려운 기능이라고 많이 하는데 처음 작성하시는거라고는 믿을 수 없을 정도로 깔끔하게 짜주셨어요 💯 

리뷰 요구사항에 남겨주신 부분은 제 의견을 한번 남겨보도록 할게요!

> final 키워드 때문에 생성자 주입 방식으로 작성하였는데 만약 통일성을 해치거나 한다면 변경하겠습니다!

`final` 키워드를 통해 불변을 선언하고 컴파일 단계에서 의도치 않은 변경을 막는 부분은 너무 좋았습니다! 저도 개발할때는 필요가 없는 부분이더라도 의도적으로 final 키워드를 선언하곤 해요. 

지금은 객체의 필드값을 final 키워드를 사용하셨는데 메서드 인자나 메서드 내의 로컬변수에도 키워드를 적극적으로 사용해도 좋을 것 같아요!

> 레포지토리를 포크해서 사용하고 있는데, 필요없는 커밋 단계가 늘어난다거나 싶으시면 말씀해주세요!

포크해서 사용하셔도 좋아요! 특히 포크해서 사용하면 서진님이 작업하신걸 개인 레포에 남겨서 정리해서 포트폴리오로 사용하기도 용이할 것 같아요!

하지만 일반적으로 실무에서는 개인 레포에 작업하지 않고 직접 조직의 레포에서 작업하는 경우가 십중 팔구 일 것이라고 생각해요! 

만약 그런 의도가 없이 단순히 익숙해서 개인 레포에 작업하시는거라면, 조직 레포에서 피쳐 브랜치를 파서 작업하는 방식이 조금더 실무맛(?) 협업을 느낄 수 있을 것 같아서 참고해서 작업 방식을 선택하시면 좋을 것 같아요

저는 지금의 방식도 확인하기 불편함 없이 너무 좋아요 👍 

> 테스트 작성이 미숙하여 오류가 많을 수 있습니다..

테스트 너무 잘 짜주셨어요! 세부 코멘트에서도 남겼지만 특히 테스트하기 어려운 부분을 하느라 정말 고생 많으셨을 것 같아요!

현재 작업해주신 부분은 실제 AWS에 요청을 보내서 동작을 검증하기 때문에 단위 테스트라기보단 통합 테스트에 가까울 것 같아요!

통합테스트의 이점은 실제 동작 그대로를 검증하기 때문에 테스트 작성 흐름이 더 자연스럽고 예기치 못한 오류를 잡을 수 있는 점이에요.
하지만, 실제 테스트코드 실행에 스프링 컨테이너를 실행해야 하기 때문에, 테스트 작업에 비용이 들게되고 시간이 오래 걸리기 때문에 개발 과정에서 잦은 테스트 실행을 막게 되요.
또 실제 외부에 의존하고 있는 부분이 있기 때문에,  만약 AWS의 서버가 잠깐 불안정해서 연결이 안된다면 똑같은 테스트코드인데도 통과하지 않는 현상이 생길 수 있어요.

그래서 그런 부분을 막고 도메인 로직을 테스트하는데 집중하시는게 더 좋아요!
지금은 단순히 파일 업로드 및 삭제 기능을 하지만 이를 더 발전시켜서 파일에 대한 데이터를 데이터베이스에 저장해야할 필요가 생기게 될 것 같아요. 이럴 떄에 도메인 로직을 단위 테스트로 검증하고, 이 도메인에서 해당 S3Uploader를 호출하는 식으로 하게되면 단위 테스트의 참맛을 더 느낄 수 있을거에요!

우선 지금은 현재 구현으로도 충분해요! 하지만 나중에 File 도메인을 만들어서 데이터베이스에 저장이 필요하다는 점! 생각해주세요!!

결론적으로는... 너무 고생 많으셨습니다! 지금 너무 잘 구현해주셔서 우선 승인은 해놓을게요! 더 개선하고 싶으신 부분은 적용하셔도 되고 아니라면 바로 머지하셔도 좋아요 👍 

고생많으셨습니다! LGTM

c0smosaur commented:
친절하고 자세한 코드 리뷰 너무 감사합니다..! 🥰 

- final 키워드를 더 적극적으로 사용해보겠습니다! 
- 레포지토리를 포크해서 사용하는 게 익숙해서 이번에도 그렇게 사용하고 있었네요..! 
말씀해주신 대로 조직 레포지토리에서 작업하는 방식으로 변경하도록 하겠습니다!
- 외부 기능보다는 도메인 로직 테스트에 더 신경쓰고,
추가로 알려주신 라이브러리나 차이점에 대해서도 더 공부하여 적용할 수 있도록 하겠습니다 🤓

일부 수정사항(`S3Config`, `S3Uploader`, 임시 `logback-test.xml`제거 등) 반영하고 머지하도록 하겠습니다!

c0smosaur committed:
refactor: 클래스명 변경 (S3Uploader -> S3FileManager)

c0smosaur committed:
remove: 임시 logback-test.xml 파일 제거

c0smosaur committed:
refactor: S3Config 수정

- @Value 애너테이션 제거
- AwsS3Properties 클래스로 속성값 주입하도록 변경
- main application에 @EnableConfigurationProperties 애너테이션 추가
- application.yml 속성값 이름 변경(변수명이 static 키워드와 중복되므로 static-region으로 변경)

c0smosaur committed:
refactor: S3FileManager 수정

- Mock Injection 수정

stoneHee99 commented:
이렇게 빠른 반영을 해주시다니 너무 감사드립니다~!

반영된 부분 너무 좋아요~! 리뷰 해드린 부분이 어떻게 보면 서진님이 느끼시기에 부담스럽도록 많거나 적용하기 힘드신 부분일 수도 있었는데 이렇게 열린 마음으로 적용해주시니 함께 만들어나가는 느낌이 들어서 저도 더 열심히 리뷰하고 또 작업할 수 있게 되는 것 같아요 다시한번 감사드립니다 :)

특히 반영된 부분 중 BDDMockito를 사용해서 given when then을 코드 표현으로 표시한 부분이 인상적이에요~! 너무 좋습니다 👍 💯 

작업해 주신 부분은 바로 머지하도록 할게요~ 고생하셨습니다! 🥇 🥇 🥇
